text/0000-modular-crypto-service.md

```diff
@@ -0,0 +1,164 @@
+title: RFC Template
+nav_order: 3
+---
+
+- Feature Name: modular_crypto_service
+- Start Date: 2020-09-19
+- RFC PR: (leave this empty)
+- Fabric Component: core, fabric-sdk-*, fabric-ca
+- Fabric Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+One paragraph explanation of the feature.
+
+This RFC aims to provide broader crypto service configurability or pluggable capability. 
+It mainly changes design of crypt/X509 usage and hash algorithm opt-in in several layer.  
+
+# Motivation
+[motivation]: #motivation
+
+Why are we doing this? What use cases does it support? What is the expected
+outcome?
+
+Following is the original initiative post, translated from Chinese.
+```
+The Hyperledger Fabric code project has a very wide range of adoptions in many domestic industries and institutions, and has a very basic role in the application of the blockchain industry.
+According to Chinese policy requirements, it is necessary to support the national secret (GM) encryption algorithm. Many domestic enterprises and Fabric enthusiasts have successively carried out some corresponding national secret (GM) fabric forks with outstanding achievements. However, in this matter, there is serious duplication of working hours and waste of human resources.
+Here, the Hyperledger Technical Working Group China (TWGC) proposes that the majority of Fabric enthusiasts and users gather and work together to create a standard national cryptographic project of Hyperledger Fabric, which is shared by everyone, and finally meet the needs of community maintainer request, merge into the Fabric up-stream.
+```
+We now have 3 streams of China crypto libraries under Hyperledger-TWGC github organization. 
+Some of them has also provided successful ane enterprise-proven bccsp implementations. 
+But as along as we go deeper into fabric source, we notice implementing bccsp only could not fulfill all of Chinese national crypto specification requirements.
+There are some cypro library usages outside of bccsp, such as communication protocol, X509 format conversion. 
+It relates to fabric architect design thus we need broader consensus and feedbacks from community as an RFC level change. 
+
+
+Additionally, the new design should take care of not only Chinese but for any other national crypto standards. 
+ 
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+There are 3 configurations controlling hash algorithm used in fabric, separately.
+- Fabric uses hash function in bccsp related to peer/orderer (controlling feature samples: signing)
+- Fabric uses hash function based on `crypto_config` section in organization/msp level (controlling feature samples: private data)
+- Fabric uses hash algorithm based on `HashingAlgorithm` value in channel level (controlling feature samples: block hash)
+
+Beside bccsp, a new X509 interface is introduced. Rework of usage of `crypto/X509` includes making `crypto/X509` as an implementaion of new X509 interface.    
```

Comment:
This is too low a level to address the problem and. in my opinion, is not generally feasible.